### PR TITLE
Chainnntfs: check historical data to determine if notification can be immediately dispatched

### PR DESCRIPTION
**Pull request by bryanvu**
Resolves issue #75

### CODE REVIEW

File: chainntnfs/btcdnotify/btcd.go [hunk 1]
```diff
@@ -447,6 +449,25 @@ func (b *BtcdNotifier) RegisterSpendNtfn(outpoint *wire.OutPoint) (*chainntnfs.S
...
 	b.notificationRegistry <- ntfn
 
+	// The following conditional checks to ensure that when a spend notification
+	// is registered, the output hasn't already been spent. If the output
+	// is no longer in the UTXO set, the chain will be rescanned from the point
+	// where the output was added. The rescan will dispatch the notification.
+	if txout, err := b.chainConn.GetTxOut(&outpoint.Hash, outpoint.Index, true); err != nil {
+		return nil, err
+	} else if txout == nil {
+		if transaction, err := b.chainConn.GetRawTransactionVerbose(&outpoint.Hash); err != nil {
```
Comment (Roasbeef) [inline chainntnfs/btcdnotify/btcd.go]: This block here seems to have some unnecessary nesting which I think can be eliminated. The block can be reformatted to look like the following: 
```
txout, err := b.chainConn.GetTxOut(&outpoint.Hash, outpoint.Index, true)
if err != nil {
    return nil, err
}

// Output has been spent.... 
if txout == nil {
    tx, err := b.chainConn.GetRawTransactionVerbose(&outpoint.Hash)
    if err != nil {
        return nil, err
    }

    blockHash, err := wire.NewShaHashFromStr(transaction.BlockHash)
    if err != nil {
        return nil, err
    }

    ops := []*wire.OutPoint{outpoint}
    if err := b.chainConn.Rescan(blockhash, nil, ops); err != nil {
        log message
        return nil, err
    } 
}
```

File: chainntnfs/interface_test.go [hunk 2]
```diff
@@ -412,13 +412,172 @@ func testMultiClientConfirmationNotification(miner *rpctest.Harness,
...
+	// First, let's send some coins to "ourself", obtainig a txid.
+	// We're spending from a coinbase output here, so we use the dedicated
+	// function.
+
+	txid, err := getTestTxId(miner)
+	if err != nil {
+		t.Fatalf("unable to create test tx: %v", err)
+	}
+
+	// Now generate two blocks. The transaction will be included in the
```
Comment (Roasbeef) [inline chainntnfs/interface_test.go]: The comment and the code are out of sync here, it says two blocks, but only on block is actually generated. It doesn't seem to affect the correctness of the test, so this is a minor nit.

File: chainntnfs/interface_test.go [hunk 3]
```diff
@@ -412,13 +412,172 @@ func testMultiClientConfirmationNotification(miner *rpctest.Harness,
...
+
+// Tests the case in which a spend notification is requested for a spend that
+// has already been included in a block. In this case, the spend notification
+// should be dispatched immediately.
+func testSpendBeforeNtfnRegistration(miner *rpctest.Harness,
+	notifier chainntnfs.ChainNotifier, t *testing.T) {
+
+	t.Logf("testing spend broadcast before notification registration")
+
+	//createTestSpends(miner *rpctest.Harness, t *testing.T)
```
Comment (Roasbeef) [inline chainntnfs/interface_test.go]: Looks like a left over line from an abandoned refactor, can you remove this line? Thanks!

**Comment by bryanvu** [comment]
Changes made. Good catches and thanks for the feedback!

**Pull request merged**